reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,807 @@
+
+Virtual clusters can be nested and a Ray job can create sub-clusters to isolate separate parts of its application workload. For example, a Tune grid sweep job can create a sub-cluster for each of its nested Train workload. These possibly nested virtual clusters form a tree where the root is the entire Ray cluster.
+
+<img src="https://user-images.githubusercontent.com/898023/291139618-0be11470-db09-466d-8c2c-37b9e5b3765c.png" alt="drawing" width="508"/>
+
+Virtual clusters with different min and max resources are autoscalable. When scaling up, virtual clusters will try to borrow more resources from their parent clusters. If their parents have available resources, the scaling up is instant. Otherwise, parents will try to borrow resources from their parents recursively and eventually this may cause the upscale of the physical cluster. When scaling down, virtual clusters will return resources to their parents recursively and eventually this may cause the downscale of the physical cluster. To ensure fairness and avoid starvation, the borrowed resources are associated with leases. Once leases expire and the tasks or actors that are currently using these resources finish, the borrowed resources will be returned back to their parent clusters so that parent clusters can lend these resources to potentially other child clusters that also need to be upscaled.
+
+Virtual nodes of a virtual cluster can be either fixed size or flexible/resizable. For a single virtual cluster, there can be multiple fixed-size virtual nodes but at most one flexible virtual node on a single physical node. The upscaling of a virtual cluster can be achieved by adding new virtual nodes or scaling up an existing flexible virtual node.
+
+### API
+
+#### Virtual Cluster Spec
+
+```
+message VirtualCluster {
+  // A virtual cluster consits of flexible resources and fixed size resources.
+
+  // == Flexible resources ==
+  // Flexible resources are stored in flexible virtual nodes
+  // and Ray will use as few flexible virtual nodes as possible
+  // for the given amount of flexible resources to minimize fragmentation.
+  // In other words, Ray prefers to scale up a flexible virtual node
+  // than adding a new flexible virtual node.
+
+  // If specified, ensure we have at least this min amount
+  // of resources before starting the cluster.
+  // If not specified, the default value is 0.
+  map<string, double> flex_resource_min
```

Comment:
Can virtual nodes use custom resources on physical nodes in addition to CPU/GPU resources?